### PR TITLE
[MRESOLVER-382] allow setting outgoing interface

### PR DESCRIPTION
**Pull request by bmarwell**
fixes MRESOLVER-382

### CODE REVIEW

File: maven-resolver-transport-http/src/main/java/org/eclipse/aether/transport/http/HttpTransporter.java [hunk 1]
```diff
@@ -295,6 +300,20 @@ final class HttpTransporter extends AbstractTransporter {
...
+    private InetAddress getBindAddress(RepositorySystemSession session) {
+        String bindAddress = ConfigUtils.getString(session, null, BIND_ADDRESS);
+        return Optional.ofNullable(bindAddress).map(this::resolveAddressOrNull).orElse(null);
+    }
+
+    private InetAddress resolveAddressOrNull(String hostOrIp) {
+        try {
+            return InetAddress.getByName(hostOrIp);
+        } catch (UnknownHostException uhe) {
+            LOGGER.warn("Given bind host (" + hostOrIp + ") cannot be resolved. Reverting to default route.");
```
Comment (michael-o) [inline maven-resolver-transport-http/src/main/java/org/eclipse/aether/transport/http/HttpTransporter.java:312]: I'd use `address` throughout and consistently...

**Review by michael-o — CHANGES_REQUESTED** [review]

File: maven-resolver-transport-http/src/main/java/org/eclipse/aether/transport/http/HttpTransporter.java (hunk 1 above)
Comment (bmarwell) [reply maven-resolver-transport-http/src/main/java/org/eclipse/aether/transport/http/HttpTransporter.java]: Thanks!
I just saw I swallowed the UHE. Maybe add a debug logging?

File: maven-resolver-transport-http/src/main/java/org/eclipse/aether/transport/http/HttpTransporter.java [hunk 2]
```diff
@@ -295,6 +300,20 @@ final class HttpTransporter extends AbstractTransporter {
...
+    private InetAddress getBindAddress(RepositorySystemSession session) {
+        String bindAddress = ConfigUtils.getString(session, null, BIND_ADDRESS);
+        return Optional.ofNullable(bindAddress).map(this::resolveAddressOrNull).orElse(null);
+    }
+
+    private InetAddress resolveAddressOrNull(String bindAddress) {
+        try {
+            return InetAddress.getByName(bindAddress);
+        } catch (UnknownHostException uhe) {
+            LOGGER.warn("Given bind address (" + bindAddress + ") cannot be resolved. Reverting to default route.");
```
Comment (michael-o) [inline maven-resolver-transport-http/src/main/java/org/eclipse/aether/transport/http/HttpTransporter.java:312]: Shouldn't this really fail with IAE? I mean, if you provide invalid values the system should fail no? If you want default, don't provide it.

Comment (bmarwell) [reply maven-resolver-transport-http/src/main/java/org/eclipse/aether/transport/http/HttpTransporter.java:312]: Yes, that would be less surprising. I agree. Will change this, too. This also answers my question about the swallowed log message.

File: maven-resolver-transport-http/src/main/java/org/eclipse/aether/transport/http/HttpTransporter.java [hunk 3]
```diff
@@ -295,6 +300,20 @@ final class HttpTransporter extends AbstractTransporter {
         this.client = builder.build();
     }
 
+    private InetAddress getBindAddress(RepositorySystemSession session) {
+        String bindAddress = ConfigUtils.getString(session, null, BIND_ADDRESS);
+        return Optional.ofNullable(bindAddress).map(this::resolveAddressOrNull).orElse(null);
+    }
+
+    private InetAddress resolveAddressOrNull(String bindAddress) {
```
Comment (michael-o) [inline maven-resolver-transport-http/src/main/java/org/eclipse/aether/transport/http/HttpTransporter.java:308]: It is not null anymore...

**Review by michael-o — APPROVED** [review]
LGTM, @cstamas ?

File: maven-resolver-transport-http/src/main/java/org/eclipse/aether/transport/http/HttpTransporter.java [hunk 4]
```diff
@@ -295,6 +300,20 @@ final class HttpTransporter extends AbstractTransporter {
...
+        String bindAddress = ConfigUtils.getString(session, null, BIND_ADDRESS);
+        return Optional.ofNullable(bindAddress).map(this::resolveAddress).orElse(null);
+    }
+
+    private InetAddress resolveAddress(String bindAddress) {
+        try {
+            return InetAddress.getByName(bindAddress);
+        } catch (UnknownHostException uhe) {
+            throw new IllegalArgumentException(
+                    "Given bind address (" + bindAddress + ") cannot be resolved. Reverting to default route.", uhe);
```
Comment (michael-o) [inline maven-resolver-transport-http/src/main/java/org/eclipse/aether/transport/http/HttpTransporter.java:313]: The message is not correct anymore

**Review by michael-o — CHANGES_REQUESTED** [review]

**Review by michael-o — APPROVED** [review]

**Review by cstamas — APPROVED** [review]
Lgtm, but a nitpick: is Optional really needed?

**Comment by bmarwell** [comment]
> Lgtm, but a nitpick: is Optional really needed?

Of course not. I just find it easier to read and write, but YMMV and I can change it to sth like:

```
if (bindAddress == null) {
  return null;
}

return this.resolveAddress(bindAddress);
```

File: maven-resolver-transport-http/src/main/java/org/eclipse/aether/transport/http/HttpTransporter.java [hunk 5]
```diff
@@ -295,6 +300,19 @@ final class HttpTransporter extends AbstractTransporter {
         this.client = builder.build();
     }
 
+    private InetAddress getBindAddress(RepositorySystemSession session) {
+        String bindAddress = ConfigUtils.getString(session, null, BIND_ADDRESS);
```
Comment (cstamas) [inline maven-resolver-transport-http/src/main/java/org/eclipse/aether/transport/http/HttpTransporter.java:304]: I can envision some show that may be "segmented" (network wise). Hence, this seems to me low hanging fruit to all o"route per remote repository", so why not?

**Comment by jira-importer** [comment]
Resolve #1061

**Pull request merged**